src/java.base/share/classes/sun/security/ssl/X509TrustManagerImpl.java

```diff
@@ -428,8 +428,17 @@ static void checkIdentity(SSLSession session,
         }
 
         if (!identifiable) {
-            checkIdentity(peerHost,
-                    trustedChain[0], algorithm, chainsToPublicCA);
+            try {
+                checkIdentity(peerHost,
+                        trustedChain[0], algorithm, chainsToPublicCA);
+            } catch(CertificateException ce) {
+                if(checkClientTrusted && "HTTPS".equalsIgnoreCase(algorithm)) {
```

Comment:
I'm not so sure this is the best fix. It seems that this should be checked earlier when the `setEndpointIdentification` method is called with a value that doesn't make sense for a client. This is not specific to your fix, but I wonder if we should go through all of the effort to create a TLS session and validate the certificate chain only to have it fail at this point. We already knew this was going to be a problem before we started the TLS session. @bradfordwetmore - can you please also weigh in on this fix?